larpandoracontent/LArTrackShowerId/SvmPfoCharacterisationAlgorithm.cc

```diff
@@ -38,41 +40,76 @@ bool SvmPfoCharacterisationAlgorithm::IsClearTrack(const Cluster *const pCluster
 
     SupportVectorMachine::DoubleVector featureVector(LArSvmHelper::CalculateFeatures(m_featureToolVector, this, pCluster));
 
-    if (m_ratioVariables)
+    if (m_trainingSetMode)
     {
-        // TODO This assumption is very bad - remove
-        const double straightLineLength(featureVector.at(0));
+        bool isTrueTrack(false);
 
-        if (straightLineLength > std::numeric_limits<double>::epsilon())
+        try
         {
-            for (unsigned int i = 1; i < featureVector.size(); ++i)
-                featureVector[i] /= straightLineLength;
+            const MCParticle *const pMCParticle(MCParticleHelper::GetMainMCParticle(pCluster));
+            isTrueTrack = ((PHOTON != pMCParticle->GetParticleId()) && (E_MINUS != std::abs(pMCParticle->GetParticleId())));
         }
+        catch (const StatusCodeException &) {}
+
+        LArSvmHelper::ProduceTrainingExample(m_trainingOutputFile, isTrueTrack, featureVector);
+        return isTrueTrack;
     }
 
+    if (!m_enableProbability)
+    {
+        return LArSvmHelper::Classify(m_supportVectorMachine, featureVector);
+    }
+    else
+    {
+        return (LArSvmHelper::CalculateProbability(m_supportVectorMachine, featureVector) > m_minProbabilityCut);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool SvmPfoCharacterisationAlgorithm::IsClearTrack(const pandora::ParticleFlowObject *const pPfo) const
+{
+    //charge related features are only calculated using hits in W view
+    ClusterList wClusterList;
+    LArPfoHelper::GetClusters(pPfo, TPC_VIEW_W, wClusterList);
+
+    PfoCharacterisationFeatureTool::FeatureToolVector featureToolVector(wClusterList.empty() ? m_featureToolVectorMissingW : m_featureToolVectorThreeD);
+    SupportVectorMachine::DoubleVector featureVector(LArSvmHelper::CalculateFeatures(featureToolVector, this, pPfo));
+    SupportVectorMachine supportVectorMachine(wClusterList.empty() ? m_supportVectorMachineMissingW : m_supportVectorMachine);
+
     if (m_trainingSetMode)
     {
         bool isTrueTrack(false);
+        int trueMCPdg(0);
 
         try
         {
-            const MCParticle *const pMCParticle(MCParticleHelper::GetMainMCParticle(pCluster));
+            const MCParticle *const pMCParticle(LArMCParticleHelper::GetMainMCParticle(pPfo));
```

Comment:
that would be something useful, in fact I realized of the 20% of cases in which it will fail (as discussed, also with Jack, in a group meeting). However in my case here I am not interested in having something else, I'd rather not consider those cases with PFOs mixed of different MC for the training examples (the purpose of using this), that's why I added a check so that if this fails I won't include that PFO in the training file